external/avro/src/main/scala/org/apache/spark/sql/avro/SparkAvroDatumReader.scala

```diff
@@ -809,51 +735,111 @@ class SparkAvroDatumReader[T](
-    override def setDecimal(ordinal: Int, value: Decimal): Unit =
-      row.setDecimal(ordinal, value, value.precision)
-  }
-
-  final class ArrayDataUpdater(var array: ArrayData)
-      extends CatalystDataUpdater {
-    override def set(ordinal: Int, value: Any): Unit = array.update(ordinal, value)
-
-    override def setNullAt(ordinal: Int): Unit = array.setNullAt(ordinal)
-    override def setBoolean(ordinal: Int, value: Boolean): Unit = array.setBoolean(ordinal, value)
-    override def setByte(ordinal: Int, value: Byte): Unit = array.setByte(ordinal, value)
-    override def setShort(ordinal: Int, value: Short): Unit = array.setShort(ordinal, value)
-    override def setInt(ordinal: Int, value: Int): Unit = array.setInt(ordinal, value)
-    override def setLong(ordinal: Int, value: Long): Unit = array.setLong(ordinal, value)
-    override def setDouble(ordinal: Int, value: Double): Unit = array.setDouble(ordinal, value)
-    override def setFloat(ordinal: Int, value: Float): Unit = array.setFloat(ordinal, value)
-    override def setDecimal(ordinal: Int, value: Decimal): Unit = array.update(ordinal, value)
+  /** Function for skipping */
+
+  // Note: Currently GenericDatumReader.skip is not working as expected, eg reading an
+  // INT for an ENUM type and not reading NULL (the later has been fixed in
+  // https://github.com/apache/avro/commit/d3a2b149aa92608956fb0b163eb1bea06ef2c05a)
+  def makeSkip(schema: Schema): ResolvingDecoder => Unit = schema.getType match {
+    case RECORD =>
+      val fieldSkips = schema.getFields.asScala.map(field => makeSkip(field.schema))
+      in => fieldSkips.foreach(f => f(in))
+
+    case ARRAY =>
+      val elementSkip = makeSkip(schema.getElementType)
+      in => {
+        var l = in.skipArray()
+        while (l > 0) {
+          var i = 0
+          while (i < l) {
+            elementSkip(in)
+            i += 1
+          }
+          l = in.skipArray()
+        }
+      }
+
+    case MAP =>
+      val valueSkip = makeSkip(schema.getValueType)
+      in => {
+        var l = in.skipMap()
+        while (l > 0) {
+          var i = 0
+          while (i < l) {
+            in.skipString()
+            valueSkip(in)
+            i += 1
+          }
+          l = in.skipMap()
+        }
+      }
+
+    case UNION =>
+      val typeSkips = schema.getTypes.asScala.map(typ => makeSkip(typ))
+      in => typeSkips(in.readIndex)(in)
+
+    case FIXED =>
+      val fixedSize = schema.getFixedSize
+      in => in.skipFixed(fixedSize)
+
+    case ENUM => in => in.readEnum()
+    case STRING => in => in.skipString()
+    case BYTES => in => in.skipBytes()
+    case INT => in => in.readInt()
+    case LONG => in => in.readLong()
+    case FLOAT => in => in.readFloat()
+    case DOUBLE => in => in.readDouble()
+    case BOOLEAN => in => in.readBoolean()
+    case NULL => in => in.readNull()
+    case _ =>
+      throw new RuntimeException("Unknown type: " + schema)
   }
 }
```

Comment:
The skip function in GenericDatumReader misses a couple of cases. So wrote this instead. Benefits of skipping can be seen in the Single Column Scan cases of the Read benchmarks (pg 2 column K of the benchmark pdf) as well as in the Pruning benchmark (pg 4 column K of the pdf).

Reply:
Did not have the filters pushdown cases in the AvroReadBenchmark in branch-3.0, but running on master I did see some improvements there as well.